crates/proto/src/xfer/dnssec_dns_handle.rs

```diff
@@ -1022,6 +1043,726 @@ fn current_time() -> u32 {
+                }
+            }
+
+            // Case 4:
+            // Name is covered by wildcard that doesn't have a record of this type
+            None => {
+                let ClosestEncloserProofInfo {
+                    closest_encloser,
+                    next_closer,
+                    closest_encloser_wildcard,
+                } = wildcard_based_encloser_proof(query_name, soa_name, nsec3s);
+                match (closest_encloser, next_closer, closest_encloser_wildcard) {
+                    (Some(_), Some(_), Some(_)) => Proof::Secure,
+                    (None, Some(_), Some(_)) if &query_name.base_name() == soa_name => {
+                        Proof::Secure
+                    }
+                    (None, None, None) if query_name == soa_name => Proof::Secure,
+                    _ => Proof::Bogus,
+                }
+            }
+        }
+    }
+
+    // Unlick non-wildcard version this cannot produce the early `Proof``
+    fn wildcard_based_encloser_proof<'a>(
+        query_name: &Name,
+        soa_name: &Name,
+        nsec3s: &'a [Nsec3RecordPair<'a>],
+    ) -> ClosestEncloserProofInfo<'a> {
+        let salt = nsec3s[0].nsec3_data.salt();
+        let iterations = nsec3s[0].nsec3_data.iterations();
+
+        let mut closest_encloser_candidates =
+            build_encloser_candidates_list(query_name, soa_name, salt, iterations);
+
+        let mut wildcard_encloser_candidates = closest_encloser_candidates
+            .iter()
+            .filter(|HashedNameInfo { name, .. }| name != soa_name)
+            .map(|info| {
+                let wildcard = info.name.clone().into_wildcard();
+                let hashed_name = nsec3hash(&wildcard, salt, iterations);
+                let base32_hashed_name = data_encoding::BASE32_DNSSEC.encode(&hashed_name);
+                HashedNameInfo {
+                    name: wildcard,
+                    hashed_name,
+                    base32_hashed_name,
+                }
+            })
+            .collect::<Vec<_>>();
+
+        let wildcard_encloser = wildcard_encloser_candidates
+            .iter()
+            .enumerate()
+            .find_map(|(index, wildcard)| {
+                let wildcard_nsec3 = nsec3s.iter().find(|record| {
+                    record.base32_hashed_name == wildcard.base32_hashed_name.as_bytes()
+                });
+                wildcard_nsec3.map(|record| (index, record))
+            })
+            .map(|(index, record)| {
+                let wildcard_name_info = wildcard_encloser_candidates.swap_remove(index);
+                (wildcard_name_info, record)
+            });
+
+        if let Some((ref wildcard_encloser_name_info, _)) = wildcard_encloser {
+            // Wildcard record exists. Within the wildcard there should be
+            // a next_closer with a name <unknown>.<wildcard_base_name>.
+
+            let closest_encloser_name = wildcard_encloser_name_info.name.base_name();
```

Comment:
I think the variable names here are pretty verbose, making it harder to read the code.

Reply:
To be fair, that's the name used in the RFC to refer to it. I guess `closest_encloser` would be an OK name but it would be confusing if there's a record for the closest encloser as well.